pycrires/pipeline.py

```diff
@@ -3884,6 +3884,140 @@ def correct_wavelengths(
+                slit_fraction_pixels = np.mean(tw_data['Upper']-tw_data['Lower'])
+                pixel_scale = full_slit_length * tot_slit_fraction  / slit_fraction_pixels
+                extraction_length = pixel_scale
+
+            # Calculate the spatial size for the extraction and
+            # the number of spatial points
+            extraction_fraction = extraction_length / full_slit_length
+            N_points = spatial_oversampling * tot_slit_fraction / extraction_fraction
+
+            # Round N_points up to the next odd integer
+            N_points = int(np.ceil(N_points) // 2 * 2 + 1)
+            
+
+            # Create EsoRex configuration file if not found
+            self._create_config("cr2res_util_extract", "util_extract", verbose)
+            config_file = self.config_folder / "util_extract.rc"
+
+            # Calculate slit fraction centers to use for util_extract
+            extraction_centers = np.linspace(0,
+                                             tot_slit_fraction - extraction_fraction,
+                                             N_points)
+            # Make sure the middle value is centered around the star
+            extraction_centers = (extraction_centers - np.median(extraction_centers)) + slit_range[1]
+            
+            # Allocate arrays to save the results
+            num_det = 3
+            results_2d = np.zeros((num_det, num_orders, N_points, 2048))
+            errors_2d = np.zeros((num_det, num_orders, N_points, 2048))
+            wavelengths = np.zeros((num_det, num_orders, 2048))
+
+            for i, center in enumerate(extraction_centers):
+                lower_lim = center - extraction_fraction/2
+                upper_lim = center + extraction_fraction/2
+                print(f'--> Extracting spectrum between slit fractions: '
+                      f'{lower_lim:.3f} - {upper_lim:.3f}')
+                
+                # Run EsoRex
+                esorex = [
+                    "esorex",
+                    f"--recipe-config={config_file}",
+                    f"--output-dir={self.product_folder}",
+                    "cr2res_util_extract",
+                    f"--slit_frac={lower_lim},{upper_lim}",
+                    f"--oversample={util_extract_oversample}",
+                    sof_file,
+                ]
+
+                if verbose:
+                    stdout = None
+                else:
+                    stdout = subprocess.DEVNULL
+                    print("Running EsoRex...", end="", flush=True)
+
+                subprocess.run(esorex, cwd=self.product_folder, stdout=stdout, check=True)
+
+                if not verbose:
+                    print(" [DONE]\n")
+                
+                # Load results
+                hdu = fits.open(self.product_folder / f"cr2res_obs_nodding_combined{nod_ab}_{count_exp:03d}_extr1D.fits")
+                for det in np.arange(1, 4):
+                    for order in np.arange(2,9):
```

Comment:
Here you could run the `for` loop over `tw_data['Order']` I think, which may then also work for other wavelengths settings where the number of orders might be different

Reply:
Done